### ISSUE TITLE
PT: Add end-to-end coverage for Authentication & 2FA flows.

### ISSUE THREAD

**Issue by lsl** [Automattic/wp-calypso#64841]
https://github.com/Automattic/wp-calypso/issues/64764 highlighted a lack of both unit and e2e test coverage for 2FA.

Add e2e test coverage for SMS+TOTP 2FA log-in flows covering routes through:
- [x] https://wordpress.com/log-in?redirect_to=https%3A%2F%2Fwordpress.com%2F (e.g. from clicking login while on wordpress.com logged out)
- [x] https://wordpress.com/log-in direct access (e.g. from google)
- [ ] Adding a Jetpack connection
- [ ] Logging into woocommerce using oauth

**Comment by worldomonation** [comment]
P2: p58i-cye-p2

**Comment by worldomonation** [comment]
@lsl I'm not as familiar with the last two use cases; could you elaborate on an example for each? Thanks!

**Comment by worldomonation** [comment]
@lsl I've been testing the behavior of TOTP 2FA and it appears that the same TOTP code is rejected the second time around even if the code is still valid.

eg. 
1. log in and enter 111 111
2. log out or launch a new browser instance
3. log in and enter 111 111 

In this case, even if 111 111 is still valid (within the 30-second window) the code is rejected.

I assume this is intended behavior, but it does introduce a problem if the TOTP spec is meant to be run concurrently against all commits.

Edit: while the [source here](https://news.ycombinator.com/item?id=30479735) is a Hacker News comment, it references the relevant parts of the specification. 

In short: yes, even if within the validity window, a used code is no longer valid.

**Comment by lsl** [comment]
> @lsl I'm not as familiar with the last two use cases; could you elaborate on an example for each? Thanks!

Just checked they're using the same form components so you can probably skip these. I was thinking about how those flows push you through a different set of urls and forms but the 2fa parts looks shared. Logging in to woocommerce.com with 2fa enabled you'll see what I mean.

> In short: yes, even if within the validity window, a used code is no longer valid.

That's going to cause problems :smile:

I'm guessing testing 2fa needs to be followed by removing 2fa so keys get refreshed.

**Comment by worldomonation** [comment]
> That's going to cause problems 😄

> I'm guessing testing 2fa needs to be followed by removing 2fa so keys get refreshed.

An alternative I'm thinking of is to run the only as part of the Pre-Release builds, which only run one at a time in sequence. That should be enough to ensure isolation and for the TOTP codes to refresh between runs.

**Comment by lsl** [comment]
That should be enough, failing there is better than nothing.

**Comment by worldomonation** [comment]
End-to-end development tasks are complete in the context of WordPress.com/Calypso.

**Comment by worldomonation** [comment]
@lsl Could you possibly elaborate on these two items?

- Adding a Jetpack connection
- Logging into woocommerce using oauth

**Comment by lsl** [comment]
The woocommerce flow is using our oauth infra on wordpress.com (mix of calypso's code and php code). 

The Jetpack connect flow is using I think part of the same oauth flow but it might be something separate, jetpack code, and probably calypso for some part of the flow.

I've also recently been made aware that woocommerce.com logins uses its own oauth code path so we might need an extra test to cover one of gravatar, akismet or pressables "login with wordpress" flow all of which use calypso and php side oauth code.

And now that I think about it, woocommerce plugin itself might have a login / connect flow that works similar to Jetpack but I'm not sure about that one.

These aren't easy tests to implement but I would call them high value integration tests.
